Add tests, add docstrings, and fix small bugs

consideRatio opened the pull request:
I've worked to understand, document, test, and fix smaller bugs I found.

### About tests

The tests/docs folder contains a sphinx project with examples used in the tests, which are overall simple should provide quite good coverage. They ensure that the sphinx directives are executed in various situations without warnings, and render .html strings of relevance.

It could have made sense to not render all the way to .html, but instead for example inspect the rST emitted by the directives somehow - but as I didn't knew how to do that I went for this end-to-end like test which probably is a good starting point to have for anyone looking to rework this further.

- Closes #17

### About small bugs

- I fixed a small edge case bug if a `:inherited-members:` options was passed to an `autoconfigurable` directive, but I don't think anyone use that as it clutters things.
  Closes #19
- The refactoring led to fixing a bug where `:exclude-members:` was bypassed, but will now be respected.
- I fixed a small bug in the README.md as introduced recently by me where I wrote about `autoclass` could be used instead of `autoconfigurable` - as it can't.
  Closes #26

### About documentation

Understanding autodoc_traits requires understanding of details of how sphinx.ext.autodoc works internally, which lacks documentation. Due to that I've invested a lot of time to help my future self and other contributors understand autodoc_traits by providing relevant references to sphinx source code and comments to summarize the important parts.

### Notes for review

I've probably invested ~16-24 hours into this. It took quite a while to onboard myself as I had to learn from comments in the sphinx source code to understand how things worked in more detail. 

I ended up adding multiple labels to this PR as it was quite extensive, which will make it show up in multiple categories if github-activity is used. A lot of the refactoring was very relevant documenting, but some refactoring also led to bugfixes etc.

consideRatio committed:
Update .gitignore

consideRatio committed:
Update MANIFEST.in to reflect single file module

consideRatio committed:
docs: fix incorrect description in README.md

consideRatio committed:
maint: make requirements sphinx>=2 traitlets>=4 explicit

consideRatio committed:
maint: add tests

consideRatio committed:
maint: refresh autodoc_traits.py for long term maintenance

minrk commented on pyproject.toml:
```diff
@@ -81,6 +83,14 @@ target_version = [
+# ref: https://docs.pytest.org/en/stable/
+#
+[tool.pytest.ini_options]
+addopts = "--verbose --color=yes --durations=10 --ignore=tests/docs/test_module.py"
```
Maybe it would be better to call it `sample_module` to avoid it being included as a test by default?

minrk commented on tests/docs/test_module.py:
```diff
@@ -0,0 +1,83 @@
+from traitlets.config.configurable import Configurable
+
+
+class TestConfigurable(Configurable):
```
Since this isn't a test module, maybe we should use `sample` naming instead of `test` naming, to avoid confusing pytest's discovery.

minrk commented on tests/README.md:
```diff
@@ -0,0 +1,30 @@
+    --watch="test_module.py" \
+    --watch="../../autodoc_traits.py" \
+    --pre-build="rm -rf build" \
+    --ignore
```
`--ignore` line wrong? It's missing a `\` continuation, but if I add it the command doesn't work. It appears to work if I remove the `--ignore`.

minrk commented on tests/docs/source/autoclass/undoc_members.rst:
```diff
@@ -0,0 +1,11 @@
+.. autoclass:: test_module.TestConfigurable
+   :noindex:
+   :members:
+   :undoc-members:
```
It doesn't seem quite right to me that adding undocumented members brings in _documented_ traits that weren't there before. It seems like if `members` only picks up non-traits by default, `undoc-members` should _also_ only pick up non-traits by default. If `undoc-members` brings in traits, then `:members:` should bring in _documented_ traits.

consideRatio replied:
Hmmm hmm but this is the `autoclass` trait.

`autoclass` will discard all members that are considered undocumented, and attributes are considered that. We rely on a hack in `autoconfigurable` to make traitlets look documented to the sphinx.ext.autodoc logic (which is inspecting `__doc__` to filter out members picked up, but considered undocumented).

Overall, I think we should not bother with the `autoclass` directive. We influence it only slightly sometimes, if its used on a class with traitlets and those traitlets are presented. Everybody should be using `autoconfigurable` for those classes.

If possible, I'd like autoclass to consider the traits with help strings documented members, but that doesn't seem possible without overriding the directive. I was trying to do that initially, but it led to sphinx warnings etc.

minrk replied:
Ah, I understand. Let's consider that a bug that we don't need to address any time soon, and open a low-priority issue for it (#30).

minrk commented on tests/docs/source/conf.py:
```diff
@@ -0,0 +1,14 @@
+import sys
+
+project = "autodoc_traits tests"
+extensions = ["autodoc_traits"]
```
adding napoleon should fix issues with reading formatted docstrings, since those packages use numpy or google-format docstrings.

```suggestion
extensions = [
    "autodoc_traits",
    # sphinx.ext.napoleon is added to avoid warnings if testing with :inherited-members:`
    # where the traitlets base classes members has numpy or google-format docstrings
    "sphinx.ext.napoleon"
]
```

minrk commented on tests/docs/source/autoconfigurable/inherited_members.rst:
```diff
@@ -0,0 +1,15 @@
+
+.. autoconfigurable:: test_module.TestConfigurableSubclass
+   :noindex:
+   :inherited-members:
```
I tried to set `:inherited-members: traitlets.HasTraits`, since the [documentation](https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html) suggests this should exclude methods inherited from this class, but it doesn't seem to work.

I've tried it on `autoclass` without autodoc-traits, and `:inherited-members: parent.Class` seems to simply not do what it says.

consideRatio replied:
I recall reading upstream sphinx source code and noting its used as a boolean flag rather than a list of members.

consideRatio replied:
HMmm, what version of sphinx was used? See https://github.com/sphinx-doc/sphinx/pull/10325 added for sphinx 5.0

minrk replied:
5.3.0. Not our problem, for now. It seems like it will be desirable _pretty often_ for documenting inheritors of HasTraits, which has _a lot_ of methods.

minrk commented on tests/docs/source/autoconfigurable/no_members.rst:
```diff
@@ -0,0 +1,10 @@
+=============================
+
+Without the ``members`` option, one may expect no members to show up, but we
+present them all still, see `Issue27`_.
```
I think it's a valid goal to present the option to export only the configurable options, and not the Python API of a class. I'm not saying `autoconfigurable` without members is necessarily the right signature, but that output should be an option.

consideRatio replied:
I agree, I commented in https://github.com/jupyterhub/autodoc-traits/issues/27#issuecomment-1351480422 about it.

minrk commented on tests/docs/source/index.rst:
```diff
@@ -0,0 +1,9 @@
+   :maxdepth: 2
+   :glob:
+
+   */*
```
I did not know about toctree globs, this is so useful!

minrk reviewed (approved):
This is truly wonderful, thanks for putting in all the effort!

Looking through the examples, I think we want to be able to support these cases:

1. document all config options on the class, but no Python and no inheritance (I think this is what `autoconfigurable` without members does now)
2. document all config options _including inherited config_, still without Python APIs (I think this is infeasible because `:inherited-members:` _also_ pulls in Python members, so you'd have to identify and manually exclude all non-configurable members, right?

Maybe the answer is to not use `:inherit-members:`, and instead define a new option like `:inherit-config-members:`. I'm not sure about needing a `:config-members:` equivalent to `:members:`, because `autoconfigurable` doesn't make any sense if you aren't going to document any configurable traits, since that's just the same output as `autoclass`, right?

I don't think you need to address these things here, because what you have is already a massive improvement. I think having these wonderful examples and illustrations merely _reveals_ that these are cases we should think about. So maybe it should be an Issue. Then again, nobody's asked for the inherit-config-but-not-python case, so maybe it doesn't matter.

minrk commented:
Actually, now that I think about it, the inherited-members being backwards means we've always been showing _inherited_ config and never showing just on-class. This is the case in jupyterhub, because [log_level is inherited from base Application](https://jupyterhub.readthedocs.io/en/stable/api/app.html#jupyterhub.app.JupyterHub.log_level) and is documented. I think that's probably what folks usually want, so maybe it is important to keep it working, and maybe even still be the default.

What do you think is the best way to signal that?

minrk commented:
Sorry for mixed messages - that doesn't mean I think it needs to block this PR, but I think it _does_ mean we need to open an Issue about it that blocks _release_ so we can make an explicit decision about it.

consideRatio committed:
refactor: rename test_module.py and classes to sample_module.py etc

consideRatio committed:
maint: add napoleon extension to test config to avoid warnings

Co-authored-by: Min RK <benjaminrk@gmail.com>

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

consideRatio committed:
docs: fix tests/README.md on how to use sphinx-autobuild

consideRatio committed:
maint: test and clarify outcome with/without members using subclass

The current behavior before and after the PR adding this commit, is that
also the inherited configurable traitlets will be presented. This commit
make us test against a class with inherited configurable traitlets to
clarify that.

consideRatio committed:
maint: rely on sphinx.ext.napoleon in tests

consideRatio commented:
@minrk I've addressed all comments, either by adding commits or commenting. On my end, this is ready for a merge but are pinging you to re-review added commits.

minrk commented on tests/docs/source/autoconfigurable/specified_members.rst:
```diff
@@ -0,0 +1,15 @@
+
+.. autoconfigurable:: sample_module.SampleConfigurable
+   :noindex:
+   :members: method,trait_nohelp
```
`trait_nohelp` is not included. Does the test cover this?

![Screenshot 2022-12-15 at 12 54 26](https://user-images.githubusercontent.com/151929/207852952-4011459c-65f2-40c9-a4e1-b21dd6e2b7d9.png)

consideRatio replied:
Hmmm... It is for me.

Here is the segment of code testing that it is: https://github.com/consideRatio/autodoc-traits/blob/pr/refresh-and-add-tests/tests/test_autodoc_traits.py#L76-L85

Here is a screenshot from my computer using sphinx 4, tested to do the same with sphinx 5 and sphinx 6beta3 also: 
![image](https://user-images.githubusercontent.com/3837114/207861147-b58b0b98-0ded-4df4-9256-1f34b5ed0f53.png)

minrk replied:
Huh, ok. Must be something weird in my setup.

minrk commented:
This is terrific, great work!

consideRatio commented:
Thanks for reviewing this @minrk!!!
